modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java

```diff
@@ -1956,73 +2062,107 @@ private void _notifySubscribers(
+			_getSubject(action, kbGroupServiceConfiguration));
 
-		subscriptionSender.addAssetEntryPersistedSubscribers(
-			KBArticle.class.getName(), kbArticle.getResourcePrimKey());
-		subscriptionSender.addPersistedSubscribers(
-			KBArticle.class.getName(), kbArticle.getGroupId());
-		subscriptionSender.addPersistedSubscribers(
-			KBArticle.class.getName(), kbArticle.getResourcePrimKey());
+		if (receivers.contains(_NOTIFICATION_RECEIVER_SUBSCRIBER)) {
+			subscriptionSender.addAssetEntryPersistedSubscribers(
+				KBArticle.class.getName(), kbArticle.getResourcePrimKey());
+			subscriptionSender.addPersistedSubscribers(
+				KBArticle.class.getName(), kbArticle.getGroupId());
+			subscriptionSender.addPersistedSubscribers(
+				KBArticle.class.getName(), kbArticle.getResourcePrimKey());
 
-		while (!kbArticle.isRoot() &&
-			   (kbArticle.getClassNameId() ==
-				   kbArticle.getParentResourceClassNameId())) {
+			KBArticle parentKBArticle = kbArticle.getParentKBArticle();
 
-			kbArticle = getLatestKBArticle(
-				kbArticle.getParentResourcePrimKey(),
-				WorkflowConstants.STATUS_APPROVED);
+			while (parentKBArticle != null) {
+				subscriptionSender.addPersistedSubscribers(
+					KBArticle.class.getName(),
+					parentKBArticle.getResourcePrimKey());
 
-			subscriptionSender.addPersistedSubscribers(
-				KBArticle.class.getName(), kbArticle.getResourcePrimKey());
+				parentKBArticle = parentKBArticle.getParentKBArticle();
+			}
+		}
+
+		if (receivers.contains(_NOTIFICATION_RECEIVER_OWNER)) {
+			User user = _userLocalService.fetchUser(kbArticle.getUserId());
+
+			if ((user == null) || !user.isActive()) {
+				user = _userLocalService.fetchUser(userId);
+			}
+
+			subscriptionSender.addRuntimeSubscribers(
+				user.getEmailAddress(), user.getFullName());
+			subscriptionSender.setSendToCurrentUser(true);
 		}
 
 		subscriptionSender.flushNotificationsAsync();
 	}
 
+	private void _notifyReviewKBArticlesByCompany(
+			Company company, Date reviewDate)
+		throws PortalException {
+
+		long userId = _userLocalService.getDefaultUserId(
+			company.getCompanyId());
+
+		List<KBArticle> kbArticles = _getKBArticlesByCompanyIdAndReviewDate(
+			company.getCompanyId(), _previousCheckDate, reviewDate);
+
+		for (KBArticle kbArticle : kbArticles) {
+			if (_log.isDebugEnabled()) {
+				_log.debug(
+					StringBundler.concat(
+						"Sending review notification for article ",
+						kbArticle.getKbArticleId(), " with reviewDate ",
+						kbArticle.getReviewDate()));
+			}
+
+			ServiceContext serviceContext = new ServiceContext();
+
+			serviceContext.setPlid(LayoutConstants.DEFAULT_PLID);
+			serviceContext.setPortalURL(
+				company.getPortalURL(kbArticle.getGroupId()));
+			serviceContext.setPortletId(KBPortletKeys.KNOWLEDGE_BASE_ADMIN);
+			serviceContext.setScopeGroupId(kbArticle.getGroupId());
+
+			_notify(
+				new HashSet<>(
+					Arrays.asList(
+						_NOTIFICATION_RECEIVER_OWNER,
+						_NOTIFICATION_RECEIVER_SUBSCRIBER)),
```

Comment:
Can you use `SetUtil.fromArray` instead?

Reply:
Yes thanks. I didn't know about it.